File: RxUI/src/main/java/com/tamsiree/rxui/view/progressing/style/Circle.java
package com.tamsiree.rxui.view.progressing.style;

import android.animation.ValueAnimator;

import com.tamsiree.rxui.view.progressing.animation.SpriteAnimatorBuilder;
import com.tamsiree.rxui.view.progressing.sprite.CircleLayoutContainer;
import com.tamsiree.rxui.view.progressing.sprite.CircleSprite;
import com.tamsiree.rxui.view.progressing.sprite.Sprite;


/**
 * @author tamsiree
 */
public class Circle extends CircleLayoutContainer {

    @Override
    public Sprite[] onCreateChild() {
        Dot[] dots = new Dot[12];
        for (int i = 0; i < dots.length; i++) {
            dots[i] = new Dot();
            dots[i].setAnimationDelay(1200 / 12 * i + -1200);
        }
        return dots;
    }

    private class Dot extends CircleSprite {

        Dot() {
            setScale(0f);
        }

        @Override
        public ValueAnimator onCreateAnimation() {
            float fractions[] = new float[]{0f, 0.5f, 1f};
            return new SpriteAnimatorBuilder(this).
                    scale(fractions, 0f, 1f, 0f).
                    duration(1200).
                    easeInOut(fractions)
                    .build();
        }
    }
}

File: RxUI/src/main/java/com/tamsiree/rxui/view/progressing/style/Circle.kt
package com.tamsiree.rxui.view.progressing.style

import android.animation.ValueAnimator
import com.tamsiree.rxui.view.progressing.animation.SpriteAnimatorBuilder
import com.tamsiree.rxui.view.progressing.sprite.CircleLayoutContainer
import com.tamsiree.rxui.view.progressing.sprite.CircleSprite
import com.tamsiree.rxui.view.progressing.sprite.Sprite

/**
 * @author tamsiree
 */
class Circle : CircleLayoutContainer() {
    override fun onCreateChild(): Array<Sprite?>? {
        val dots = arrayOfNulls<Sprite>(12)
        for (i in dots.indices) {
            dots[i] = Dot()
            dots[i]!!.setAnimationDelay(1200 / 12 * i + -1200)
        }
        return dots
    }

    private inner class Dot internal constructor() : CircleSprite() {
        override fun onCreateAnimation(): ValueAnimator? {
            val fractions = floatArrayOf(0f, 0.5f, 1f)
            return SpriteAnimatorBuilder(this).scale(fractions, 0f, 1f, 0f).duration(1200).easeInOut(*fractions)
                    .build()
        }

        init {
            setScale(0f)
        }
    }
}